File: python/calculadora.py
class Calculadora:
    """Calculadora de operações matemáticas básicas."""

    def somar(self, x: float, y: float) -> float:
        return x + y

    def subtrair(self, x: float, y: float) -> float:
        return x - y

    def multiplicar(self, x: float, y: float) -> float:
        return x * y

    def dividir(self, x: float, y: float) -> float:
        if y == 0:
            raise ZeroDivisionError("divisão por zero não é permitida")
        return x / y

    def fatorial(self, numero: int) -> int:
        if not isinstance(numero, int) or isinstance(numero, bool):
            raise TypeError("o fatorial aceita apenas números inteiros")
        if numero < 0:
            raise ValueError("não é possível calcular o fatorial de um número negativo")

        resultado = 1
        for fator in range(2, numero + 1):
            resultado *= fator
        return resultado

    def potencia(self, base: float, expoente: int) -> float:
        if base == 0 and expoente < 0:
            raise ZeroDivisionError("zero não pode ser elevado a um expoente negativo")
        return base**expoente

File: calculadora.go
package calculadora

func Soma(x, y float64) float64 {
	return x + y
}

func Subtracao(x, y float64) float64 {
	return x - y
}

func Multiplicacao(x, y float64) float64 {
	return x * y
}

func Divisao(x, y float64) float64 {
	if y == 0 {
		panic("Resultado Indefinido, divisão por zero não é permitida")
	}
	return x / y
}

func Fatorial(x int) int {
	if x < 0 {
		panic("Não é possível calcular fatorial para números negativos")
	}
	if x == 0 || x == 1 {
		return 1
	}
	return x * Fatorial(x-1)
}

func Potencia(base float64, expoente int) float64 {
	if expoente == 0 {
		return 1
	}
	if expoente < 0 {
		return 1 / Potencia(base, -expoente)
	}
	result := base
	for i := 1; i < expoente; i++ {
		result *= base
	}
	return result
}
